utils/script-luajit.c

```diff
@@ -0,0 +1,405 @@
+	dllua_settable(L, -3);
+	dllua_pushstring(L, "timestamp");
+	dllua_pushinteger(L, sc_ctx->timestamp);
+	dllua_settable(L, -3);
+	dllua_pushstring(L, "duration");
+	dllua_pushinteger(L, sc_ctx->duration);
+	dllua_settable(L, -3);
+	dllua_pushstring(L, "address");
+	dllua_pushinteger(L, sc_ctx->address);
+	dllua_settable(L, -3);
+	dllua_pushstring(L, "name");
+	dllua_pushstring(L, sc_ctx->name);
+	dllua_settable(L, -3);
+}
+
+static void setup_argument_context(bool is_retval, struct script_context *sc_ctx)
+{
+	struct uftrace_arg_spec *spec;
+	void *data = sc_ctx->argbuf;
+	union script_arg_val val;
+	int count = 0;
+
+	list_for_each_entry(spec, sc_ctx->argspec, list) {
+		/* skip unwanted arguments or retval */
+		if (is_retval != (spec->idx == RETVAL_IDX))
+			continue;
+
+		count++;
+	}
+	if (count == 0)
+		return;
+
+	if (is_retval)
+		dllua_pushstring(L, "retval");
+	else
+		dllua_pushstring(L, "args");
+	dllua_newtable(L);
+
+	count = 0;
+	list_for_each_entry(spec, sc_ctx->argspec, list) {
+		const int null_str = -1;
+		unsigned short slen;
+		char ch_str[2];
+		char *str;
+		double dval;
+
+		/* skip unwanted arguments or retval */
+		if (is_retval != (spec->idx == RETVAL_IDX))
+			continue;
+
+		/* reset the value */
+		memset(val.v, 0, sizeof(val));
+
+		switch (spec->fmt) {
+		case ARG_FMT_AUTO:
+		case ARG_FMT_SINT:
+		case ARG_FMT_UINT:
+		case ARG_FMT_HEX:
+		case ARG_FMT_PTR:
+		case ARG_FMT_ENUM:
+			memcpy(val.v, data, spec->size);
+			switch (spec->size) {
+			case 1:
+				dllua_pushinteger(L, 1 + count++);
+				dllua_pushinteger(L, val.c);
+				dllua_settable(L, -3);
+				break;
+			case 2:
+				dllua_pushinteger(L, 1 + count++);
+				dllua_pushinteger(L, val.s);
+				dllua_settable(L, -3);
+				break;
+			case 4:
+				dllua_pushinteger(L, 1 + count++);
+				dllua_pushinteger(L, val.i);
+				dllua_settable(L, -3);
+				break;
+			case 8:
+				dllua_pushinteger(L, 1 + count++);
+				dllua_pushinteger(L, val.L);
+				dllua_settable(L, -3);
+				break;
+			default:
+				pr_warn("invalid argument format: %d\n", spec->fmt);
+				break;
+			}
+			data += ALIGN(spec->size, 4);
+			break;
+		case ARG_FMT_FLOAT:
+			memcpy(val.v, data, spec->size);
+			switch (spec->size) {
+			case 4:
+				dval = val.f;
+				break;
+			case 8:
+				dval = val.d;
+				break;
+			case 10:
+				dval = (double)val.D;
+				break;
+			default:
+				pr_dbg("invalid floating-point type size %d\n",
+				       spec->size);
+				dval = 0;
+				break;
+			}
+			dllua_pushinteger(L, 1 + count++);
+			dllua_pushnumber(L, dval);
+			dllua_settable(L, -3);
+			data += ALIGN(spec->size, 4);
+			break;
+
+		case ARG_FMT_STR:
+		case ARG_FMT_STD_STRING:
+			/* get string length (2 bytes in the beginning) */
+			memcpy(&slen, data, 2);
+
+			str = xmalloc(slen + 1);
+
+			/* copy real string contents */
+			memcpy(str, data + 2, slen);
+			str[slen] = '\0';
+
+			/* NULL string is encoded as '0xffffffff' */
+			if (!memcmp(str, &null_str, sizeof(null_str)))
```

Comment:
The str might be less than 4 bytes, you need to check `slen == 4` before this.

Reply:
Fixed.